File: Workpiece/Summoner/LocalPackages/com.gamelovers.mcp-unity/Editor/Tools/UpdateGameObjectTool.cs
using System;
using UnityEngine;
using UnityEditor;
using McpUnity.Utils; // For GameObjectHierarchyCreator and McpLogger
using McpUnity.Unity; // For McpUnitySocketHandler
using Newtonsoft.Json.Linq; // For JObject

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for updating or creating a GameObject in the Unity Editor.
    /// Supports setting name, tag, layer, active state, and static state by instance ID or hierarchy path.
    /// Returns a JObject result similar to UpdateComponentTool for consistency.
    /// </summary>
    public class UpdateGameObjectTool : McpToolBase
    {
        public UpdateGameObjectTool()
        {
            Name = "update_gameobject";
            Description = "Updates or creates a GameObject and its properties (name, tag, layer, active state, static state) based on instance ID or object path.";
            IsAsync = false; // Operations are expected to be quick
        }

        /// <summary>
        /// Executes the update or creation of a GameObject based on the provided parameters.
        /// </summary>
        /// <param name="parameters">A JObject containing: instanceId (int?), objectPath (string), name (string), tag (string), layer (int?), isActiveSelf (bool?), isStatic (bool?)</param>
        /// <returns>JObject with success, message, instanceId, name, and path fields (see UpdateComponentTool for format)</returns>
        public override JObject Execute(JObject parameters)
        {
            // Extract parameters from JObject
            int? instanceId = parameters["instanceId"]?.ToObject<int?>();
            string objectPath = parameters["objectPath"]?.ToObject<string>();
            JObject gameObjectData = parameters["gameObjectData"] as JObject;

            string newName = gameObjectData? ["name"]?.ToObject<string>();
            string newTag = gameObjectData? ["tag"]?.ToObject<string>();
            int? newLayer = gameObjectData? ["layer"]?.ToObject<int?>();
            bool? newIsActiveSelf = (gameObjectData?["activeSelf"] ?? gameObjectData?["isActiveSelf"])?.ToObject<bool?>();
            bool? newIsStatic = (gameObjectData?["isStatic"] ?? gameObjectData?["static"])?.ToObject<bool?>();

            GameObject targetGameObject = null;
            string identifierInfo = "";

            // Identify or create the GameObject by instanceId or objectPath
            if (instanceId.HasValue)
            {
                targetGameObject = EditorUtility.InstanceIDToObject(instanceId.Value) as GameObject;
                identifierInfo = $"instance ID {instanceId.Value}";
            }
            else if (!string.IsNullOrEmpty(objectPath))
            {
                // Will create the GameObject if it doesn't exist
                targetGameObject = GameObjectHierarchyCreator.FindOrCreateHierarchicalGameObject(objectPath);
                identifierInfo = $"path '{objectPath}'";
            }
            else
            {
                // Neither instanceId nor objectPath was provided
                return McpUnitySocketHandler.CreateErrorResponse("Either 'instanceId' or 'objectPath' must be provided.", "validation_error");
            }

            // Check if we could not identify or create the GameObject
            if (targetGameObject == null)
            {
                return McpUnitySocketHandler.CreateErrorResponse($"Target GameObject could not be identified or created using {identifierInfo}.", "unknown_error");
            }

            // Record for undo in Unity Editor
            Undo.RecordObject(targetGameObject, "Update GameObject Properties");
            bool propertiesUpdated = false;
            string originalNameForLog = targetGameObject.name;

            // Update name if provided and different
            if (!string.IsNullOrEmpty(newName) && targetGameObject.name != newName)
            {
                targetGameObject.name = newName;
                propertiesUpdated = true;
            }

            // Update tag if provided and different, warn if tag doesn't exist
            if (!string.IsNullOrEmpty(newTag))
            {
                bool tagExists = Array.Exists(UnityEditorInternal.InternalEditorUtility.tags, t => t.Equals(newTag));
                if (!tagExists)
                {
                    McpLogger.LogWarning($"UpdateGameObjectTool: Tag '{newTag}' does not exist for GameObject '{originalNameForLog}'. Tag not changed. Please create the tag in Unity's Tag Manager.");
                }
                else if (!targetGameObject.CompareTag(newTag))
                {
                    targetGameObject.tag = newTag;
                    propertiesUpdated = true;
                }
            }

            // Update layer if provided and valid
            if (newLayer.HasValue)
            {
                if (newLayer.Value < 0 || newLayer.Value > 31)
                {
                    McpLogger.LogWarning($"UpdateGameObjectTool: Invalid layer value {newLayer.Value} for GameObject '{originalNameForLog}'. Layer must be between 0 and 31. Layer not changed.");
                }
                else if (targetGameObject.layer != newLayer.Value)
                {
                    targetGameObject.layer = newLayer.Value;
                    propertiesUpdated = true;
                }
            }

            // Update active state if provided and different
            if (newIsActiveSelf.HasValue && targetGameObject.activeSelf != newIsActiveSelf.Value)
            {
                targetGameObject.SetActive(newIsActiveSelf.Value);
                propertiesUpdated = true;
            }

            // Update static state if provided and different
            if (newIsStatic.HasValue && targetGameObject.isStatic != newIsStatic.Value)
            {
                targetGameObject.isStatic = newIsStatic.Value;
                propertiesUpdated = true;
            }

            // Mark as dirty if any property was changed
            if (propertiesUpdated)
            {
                EditorUtility.SetDirty(targetGameObject);
            }

            // Compose result message and return as JObject (like UpdateComponentTool)
            return new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = propertiesUpdated
                    ? $"GameObject '{targetGameObject.name}' (identified by {identifierInfo}) updated successfully."
                    : $"No properties were changed for GameObject '{targetGameObject.name}' (identified by {identifierInfo}).",
                ["instanceId"] = targetGameObject.GetInstanceID(),
                ["name"] = targetGameObject.name,
                ["path"] = GetGameObjectPath(targetGameObject)
            };
        }

        /// <summary>
        /// Utility to get the hierarchy path of a GameObject (root/child/.../target)
        /// </summary>
        /// <param name="obj">The GameObject to get the path for</param>
        /// <returns>Hierarchy path as string</returns>
        private static string GetGameObjectPath(GameObject obj)
        {
            if (obj == null) return null;
            string path = "/" + obj.name;
            while (obj.transform.parent != null)
            {
                obj = obj.transform.parent.gameObject;
                path = "/" + obj.name + path;
            }
            return path;
        }
    }
}

File: Workpiece/Summoner/LocalPackages/com.gamelovers.mcp-unity/Server~/src/tools/updateGameObjectTool.ts
import * as z from 'zod';
import { McpUnity } from '../unity/mcpUnity.js';
import { Logger } from '../utils/logger.js';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

// Constants for the tool
const toolName = 'update_gameobject';
const toolDescription = 'Updates properties of a GameObject in the Unity scene by its instance ID or path. If the GameObject does not exist at the specified path, it will be created.';
const paramsSchema = z.object({
  instanceId: z.number().optional().describe('The instance ID of the GameObject to update'),
  objectPath: z.string().optional().describe('The path of the GameObject in the hierarchy to update (alternative to instanceId)'),
  gameObjectData: z.object({
    name: z.string().optional().describe('New name for the GameObject'),
    tag: z.string().optional().describe('New tag for the GameObject'),
    layer: z.number().int().optional().describe('New layer for the GameObject'),
    activeSelf: z.boolean().optional().describe('Set the active state of the GameObject (GameObject.SetActive(value))'),
    isStatic: z.boolean().optional().describe('Set the static state of the GameObject (GameObject.isStatic = value)'),
  }).describe('An object containing the fields to update on the GameObject. If the GameObject does not exist at objectPath, it will be created.')
    .refine(data => Object.keys(data).length > 0, { message: 'gameObjectData must contain at least one property to update.' }),
});

/**
 * Creates and registers the Update GameObject tool with the MCP server
 * This tool allows updating or creating GameObjects into the Unity Editor active scene
 * 
 * @param server The MCP server instance to register with
 * @param mcpUnity The McpUnity instance to communicate with Unity
 * @param logger The logger instance for diagnostic information
 */
export function registerUpdateGameObjectTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
    logger.info(`Registering tool: ${toolName}`);
        
    // Register this tool with the MCP server
    server.tool(
      toolName,
      toolDescription,
      paramsSchema.shape,
      async (params: any) => {
        try {
          logger.info(`Executing tool: ${toolName}`, params);
          const result = await toolHandler(mcpUnity, params);
          logger.info(`Tool execution successful: ${toolName}`);
          return result;
        } catch (error) {
          logger.error(`Tool execution failed: ${toolName}`, error);
          throw error;
        }
      }
    );
}
/**
 * Handles updating or creating GameObjects into the Unity Editor active scene
 * 
 * @param mcpUnity The McpUnity instance to communicate with Unity
 * @param params The parameters for the tool
 * @returns A promise that resolves to the tool execution result
 * @throws McpUnityError if the request to Unity fails
 */
async function toolHandler(mcpUnity: McpUnity, params: any): Promise<CallToolResult> {
    // Validate parameters - require either instanceId or objectPath
    if ((params.instanceId === undefined || params.instanceId === null) && 
        (!params.objectPath || params.objectPath.trim() === '')) {
        throw new McpUnityError(
        ErrorType.VALIDATION,
        "Either 'instanceId' or 'objectPath' must be provided"
        );
    }

    const response = await mcpUnity.sendRequest({
        method: 'update_gameobject',
        params: {
          instanceId: params.instanceId,
          objectPath: params.objectPath,
          gameObjectData: params.gameObjectData,
        }
      });
  
      if (!response.success) {
        throw new McpUnityError(
          ErrorType.TOOL_EXECUTION,
          response.message || `Failed to update the GameObject`
        );
      }
  
      // Create a description of which GameObject was targeted
      const targetDescription = params.objectPath 
        ? `path '${params.objectPath}'` 
        : `ID ${params.instanceId}`;
      
      return {
        content: [{
          type: response.type,
          text: response.message || `Successfully updated the GameObject with ${targetDescription}`
        }]
      };
}
